### PR TITLE
Sparse implementation of `vcov_CR3J`

### PR DESCRIPTION
**Pull request by kylebutts**
1. Adds the option `sparse` to `vcov_CR3J`. This is substantially faster (see reprex), but (1) requires `fixest::sparse_model_matrix` to be merged (https://github.com/lrberge/fixest/pull/418) and (2) results in different `beta_jack`. This is fine for the vcov, but makes `summclust` not match Stata. All tests passed and I added a set of tests to make sure sparse/dense versions match each other.

2. Added a little speedup to partial-leverage. In R, you basically never want to use `solve(A) %*% b`. It's an order of magnitude slower than `Matrix::solve(A, b)`.

3. Made two modifications to existing tests. For `CR3J_vs_HC3`, I dropped `nlswork` to 100 rows. It's much faster and effectively the same test. For the plot and the summary, I added a `quietly` function so it doesn't print out a bunch of results and open the plot pane when testing.

---
`vcov_CR3J` sparse example:

``` r
library(fixest)
library(Matrix)
library(summclust)
library(devtools)
#> Loading required package: usethis
# load_all()
# devtools::check(vignettes = FALSE, env_vars = c())

nlswork <- haven::read_dta("http://www.stata-press.com/data/r9/nlswork.dta")

# drop NAs at the moment
nlswork <- nlswork[, c(
  "ln_wage", "grade", "age", "birth_yr",
  "union", "race", "msp", "ind_code"
)]
nlswork <- na.omit(nlswork)

# 30 clusters
nlswork$group <- (seq_len(nrow(nlswork)) - 1) %% 30 + 1


lm_fit <- lm(
  ln_wage ~ 0 + factor(union) + factor(msp) + factor(race),
  data = nlswork
)
feols_fit <- feols(
  ln_wage ~ 0 + i(union) + i(msp) + i(race),
  data = nlswork
)
obj <- feols_fit


vcovCR3_lm <- vcov_CR3J(lm_fit, ~group, type = "CRV3")
sqrt(diag(vcovCR3_lm))[c("factor(union)0", "factor(union)1", "factor(msp)1", "factor(race)2", "factor(race)3")]
#> factor(union)0 factor(union)1   factor(msp)1  factor(race)2  factor(race)3 
#>    0.004831951    0.006739998    0.006764424    0.006009833    0.028693133

tictoc::tic("Sparse")
vcovCR3_sparse <- vcov_CR3J(feols_fit, ~group, type = "CRV3", sparse = TRUE)
sqrt(diag(vcovCR3_sparse))
#>    union::0    union::1      msp::1     race::2     race::3 
#> 0.004831951 0.006739998 0.006764424 0.006009833 0.028693133
tictoc::toc()
#> Sparse: 0.15 sec elapsed

tictoc::tic("Dense")
vcovCR3_dense <- vcov_CR3J(feols_fit, ~group, type = "CRV3", sparse = FALSE)
sqrt(diag(vcovCR3_dense))
#>    union::0    union::1      msp::1     race::2     race::3 
#> 0.004831951 0.006739998 0.006764424 0.006009833 0.028693133
tictoc::toc()
#> Dense: 1.989 sec elapsed
```

<sup>Created on 2023-06-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

---
`solve(A) %*% b` vs. `solve(A, b)`

``` r
n <- 1000
A <- matrix(runif(n^2), nrow = n, ncol = n)
b <- runif(n)

bench::mark(
  as.numeric(solve(A) %*% b),
  as.numeric(solve(A, b)),
  iterations = 10
)
#> # A tibble: 2 × 6
#>   expression                      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                 <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 as.numeric(solve(A) %*% b)    405ms    410ms      2.42      23MB    5.65 
#> 2 as.numeric(solve(A, b))       126ms    126ms      7.88    7.67MB    0.876
```

<sup>Created on 2023-06-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Comment by s3alfisc** [comment]
Hi @kylebutts , thanks so much for this PR. The speed gains look amazing. I have taken a first look atthe PR , and I think it looks excellent, and I will be more than happy to accept this PR once the sparse model matrix makes it into fixest. 

I have a few comments, which I am listing here (but I will also comment in the PR). Some of these, I might just fix myself as I go through the code line by line - in case I do this, I will mark it as done (I hope that's ok with you?). 

- [ ] `=` assignment (as in fixest) vs `<-` assignment as in the rest of summclust
- [ ] as I wrote the other day, `arma::pinv` is around twice as fast as `MASS::ginv`. For performance reasons, it might make sense to switch, at the cost of adding `RcppArmadillo` as a (heavy) dependency. But all the Rcpp packages are extremely well maintained (practically one epsilon below base R), so I would be fine to add this dependency

Other general comments I had: 

- I like your solution of setting the `sparse = TRUE` as default for `vcov_CR3J`. If the `summclust` function has no `sparse` argument itself, users will never have the option to produce results for the leverage statistics that differ from stata (at the cost of small drops in performance). We can even consider to return slightly fewer results for the `vcov_CR3J` class when `sparse = TRUE`, as practically all results are only exported for computing leverage stats. E.g. we could add a second class attribute as 
```r
class(res) <- c("vcov_CR3J", "sparse") 
```
and then not return the beta_jack's etc for this class. 

I'll play around with the code a little bit more and might add to this list. Anyways, thanks a lot - this is a really exciting PR! =) 

Best, Alex

File: R/vcov3_fixest.R [hunk 1]
```diff
@@ -244,3 +263,113 @@ vcov_CR3J.fixest <- function(
...
+    stop(paste0(
+      "The following variables are not found in the dataset used in your `feols` call: ", 
+      paste(cluster_vars[!cluster_vars_in_data], collapse = ", ")
+    ))
+  }
+
+  # Assumes that length(cluster_vars) == 1 (for now; can modify later for multi-way clustering) 
+  if (length(cluster_vars) > 1) stop("Only 1 cluster variable supported right now")
+
+  cluster_vec = data[[cluster_vars]]
```
Comment (s3alfisc) [inline R/vcov3_fixest.R:301]: This causes some issues if there are NA variables in the cluster vector. 

E.g. if the first element of `cluster_vars` is not dropped, `unique_clusters` below will have a cluster group "NA". 

In my understanding, fixest handles this by already dropping columns from y and X if there is a NA value in the clustering variable. If the clustering variable is changed on the fly, I think that fixest throws an error. This is certainly what sandwich does. One solution here would be to enforce that 

```r
length(cluster_vec) == nrow(X)
```
and require the user to clean the clustering variable prior to using `vcov_CR3J`.

Comment (s3alfisc) [reply R/vcov3_fixest.R:301]: That's actually something I really need to fix for the current version of `summclust`, I will hopefully release a new version today.

Comment (s3alfisc) [reply R/vcov3_fixest.R:301]: [This ](https://github.com/s3alfisc/summclust/blob/0f99aaf2846f94ea67b05440528fa6ceb4d57e9a/R/get_cluster.R#L81)is how you can drop missings dropped by `fixest`. Then just throw an error if the cluster vector is still not of the same dimension as the model matrix / if it still includes errors.

Comment (kylebutts) [reply R/vcov3_fixest.R:301]: Yeah, I probably should have been more careful about that. Keeping only the rows used in the original estimation, `cluster_var[object$obs_selection]`, makes sense to me. 

I think that if there are any NAs in `cluster_var[object$obs_selection]`, then an error should be returned. I think that b/c the point estimates will change relative to the original regression and that can be confusing.

Comment (s3alfisc) [reply R/vcov3_fixest.R:301]: Yes, I think that's more or less the standard - sandwich, fwildclusterboot and fixest all do this. The released version of summclust will hopefully also do this soon 😀

**Comment by kylebutts** [comment]
`MPinv` from VCA package produces the same results as `MASS::ginv`. It passes all the tests and should work now. However, still using sparse produces different results on summclust.  

The reason I pushed onto this PR is that `vcov_CR3J.fixest` now works fully without densifying (may be a made up word) matrices

File: R/MPinv.R [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+#' @param tol (numeric) tolerance value to be used in comparisons 
+#' 
+#' @return (object) A Moore-Penrose inverse of X. 
+#' 
+#' @author Authors of the 'MASS' package.  
+MPinv <- function (X, tol = sqrt(.Machine$double.eps))  {
+  if (length(dim(X)) > 2L) 
+    stop("'X' must be two-dimensional")
+  
+  Xsvd <- svd(X)
```
Comment (s3alfisc) [inline R/MPinv.R:20]: I think that `svd()` converts sparse matrices to dense matrices under the hood? I have just done some minor performance benchmarks, and it looks to me like MPinv is just as fast as `MASS::ginv()`: 

```r
library(VCA)
library(Matrix)
library(microbenchmark)

N <- 100000
X <- factor(sample(1:1000, N, TRUE))
df <- data.frame(X = X)
mf <- model.frame(~X,df)
mm <- sparse.model.matrix(mf)
dim(mm)

XX <- t(mm) %*% mm
dim(XX)

microbenchmark(
  VCA::MPinv(XX), 
  MASS::ginv(as.matrix(XX)), 
  times = 1
)

# Unit: seconds
# expr      min       lq     mean   median       uq      max
# VCA::MPinv(XX) 5.654788 5.654788 5.654788 5.654788 5.654788 5.654788
# MASS::ginv(as.matrix(XX)) 4.829764 4.829764 4.829764 4.829764 4.829764 4.829764

```

Comment (s3alfisc) [reply R/MPinv.R:20]: What do you think of adding armadillo as a dependency? I have added it to the dev branch of `fwildclusterboot` [here](https://github.com/s3alfisc/fwildclusterboot/blob/7a00a08cd96ffb38efade7ccad47c0b2a8160d14/src/pinv.cpp) (but RccpArma is already a dependency).

Comment (s3alfisc) [reply R/MPinv.R:20]: Updated benchmarks running `fwildclusterboot:::pinv`: 

```r
# expr      min       lq     mean   median       uq
# VCA::MPinv(XX) 5.578639 5.578639 5.578639 5.578639 5.578639
# MASS::ginv(as.matrix(XX)) 5.767423 5.767423 5.767423 5.767423 5.767423
# fwildclusterboot:::pinv(as.matrix(XX)) 2.993501 2.993501 2.993501 2.993501 2.993501

```

Comment (s3alfisc) [reply R/MPinv.R:20]: Last benchmark - Eigen's pinv implementation beats Armadillo, by quite a large margin: 

```r
library(VCA)
library(Matrix)
library(microbenchmark)

N <- 100000
X <- factor(sample(1:1000, N, TRUE))
df <- data.frame(X = X)
mf <- model.frame(~X,df)
mm <- sparse.model.matrix(mf)
dim(mm)

XX <- t(mm) %*% mm
dim(XX)

#x <- sparseEigen::spEigen(XX, q = ncol(XX))


microbenchmark(
  Matrix::solve(XX),
  VCA::MPinv(XX), 
  MASS::ginv(as.matrix(XX)), 
  fwildclusterboot:::pinv(as.matrix(XX)), 
  eigenR = EigenR::Eigen_pinverse(as.matrix(XX)), 
  #sparseEigen = sparseEigen::spEigen(XX, q = ncol(XX)),
  #eigen = eigen(XX),
  times = 1
)

# Unit: milliseconds
# expr       min
# Matrix::solve(XX)   58.4333
# VCA::MPinv(XX) 6007.1988
# MASS::ginv(as.matrix(XX)) 4880.4366
# fwildclusterboot:::pinv(as.matrix(XX)) 2981.8584
# eigenR  779.7848
# lq      mean    median        uq       max neval
# 58.4333   58.4333   58.4333   58.4333   58.4333     1
# 6007.1988 6007.1988 6007.1988 6007.1988 6007.1988     1
# 4880.4366 4880.4366 4880.4366 4880.4366 4880.4366     1
# 2981.8584 2981.8584 2981.8584 2981.8584 2981.8584     1
# 779.7848  779.7848  779.7848  779.7848  779.7848     1
```

Comment (s3alfisc) [reply R/MPinv.R:20]: Implementation [here](https://github.com/stla/EigenR/blob/1ffb3091d4bfa6fb0cd5fbfdaa1cc06f7bb6ea69/src/Inverse.cpp#L39). You can potentially make it even faster by parallelizing (I do this in [fwildclusterboot](https://github.com/s3alfisc/fwildclusterboot/blob/master/src/utils.cpp)).

**Pull request merged**